### PR TITLE
Restore command state support #925

### LINKED ISSUE #925: Command handler state is not observed
This is discussed in [Bug 549802](https://bugs.eclipse.org/bugs/show_bug.cgi?id=549802)

Corresponding tests are disabled:
https://github.com/eclipse-platform/eclipse.platform.ui/blob/50553ea8426afdc390e14676abf8906c44d184a4/tests/org.eclipse.ui.tests/Eclipse%20UI%20Tests/org/eclipse/ui/tests/commands/StateTest.java#L43
https://github.com/eclipse-platform/eclipse.platform.ui/blob/50553ea8426afdc390e14676abf8906c44d184a4/tests/org.eclipse.ui.tests/Eclipse%20UI%20Tests/org/eclipse/ui/tests/commands/ToggleStateTest.java#L45

Commands try to subscribe to handler state in `org.eclipse.core.commands.Command.addState(String, State)` and `org.eclipse.core.commands.Command.setHandler(IHandler)`, but handler proxies fail to propagate `org.eclipse.core.commands.IObjectWithState` protocol.

Offending proxies:

- org.eclipse.ui.internal.WorkbenchHandlerServiceHandler
- org.eclipse.ui.internal.handlers.E4HandlerProxy

While `org.eclipse.ui.internal.handlers.HandlerProxy.handleStateChange(State, Object)` does propagate the states to the actual handler, it is never called as upstream proxies do not invoke `HandlerProxy.addState()`

[Attached is a simple plugin with a stateful command handler](https://github.com/eclipse-platform/eclipse.platform.ui/files/11995015/handler_state_fail.zip), events from which fail to propagate to command framework.
The menu item is "File/I should blink". It does not change state even if menu is reopened.

### Workaround
if command id is known, update command state through service locator:
```
	@Override
	public void setEnabled(Object evaluationContext) {
		super.setEnabled(evaluationContext);
		boolean isRegisteredIdle;
		try {
			Stream<JavaThread> threads = getThreads(evaluationContext);
			isRegisteredIdle = threads.parallel()
					.anyMatch(Activator.getDefault()::isIdle);
		} catch (ExecutionException e) {
			LOG.error("Failed to process selected threads", e);
			isRegisteredIdle = false;
		}
		
		IWorkbenchSite site;
		try {
			site = HandlerUtil.getActiveSiteChecked(adapt(evaluationContext));
		} catch (ExecutionException e) {
			throw new IllegalStateException(e);
		}
		State state = site.getService(ICommandService.class).getCommand("org.basilevs.jstackfilter.eclipse.ui.markidle").getState(RegistryToggleState.STATE_ID);
		state.setValue(isRegisteredIdle);
	}
```

### PR DESCRIPTION
**Pull request by basilevs**
Restore state support for command handlers as documented in [IObjectWithState](https://github.com/eclipse-platform/eclipse.platform.ui/blob/50553ea8426afdc390e14676abf8906c44d184a4/bundles/org.eclipse.core.commands/src/org/eclipse/core/commands/IObjectWithState.java#L29) Javadoc.
See previous work in [Bug 549802](https://bugs.eclipse.org/bugs/show_bug.cgi?id=549802)
Fixes #925

This PR assumes that `IObjectWithState` protocol is not deprecated. If it is feasible to deprecate it, do not merge this PR.

Command state support had been lost during migration to E4 as proxies failed to properly support `IObjectWithState` protocol or ignored it altogether.

- [HandlerProxy](https://github.com/eclipse-platform/eclipse.platform.ui/blob/50553ea8426afdc390e14676abf8906c44d184a4/bundles/org.eclipse.ui.workbench/Eclipse%20UI/org/eclipse/ui/internal/handlers/HandlerProxy.java#L63)
- [HandlerServiceHandler](https://github.com/eclipse-platform/eclipse.platform.ui/blob/50553ea8426afdc390e14676abf8906c44d184a4/bundles/org.eclipse.e4.core.commands/src/org/eclipse/e4/core/commands/internal/HandlerServiceHandler.java#L36)
- [E4HandlerProxy](https://github.com/eclipse-platform/eclipse.platform.ui/blob/50553ea8426afdc390e14676abf8906c44d184a4/bundles/org.eclipse.ui.workbench/Eclipse%20UI/org/eclipse/ui/internal/handlers/E4HandlerProxy.java#L53)

do now support the protocol. Following tests are modified to rely on command state polling and reenabled:

StateTest
ToggleStateTest
RadioStateTest

A new test WorkbenchStateTest demonstrates how state propagation works in workbench environment without explicit polling.

Main changes:
- internal APIs are extended to implement IObjectWIthState
- HandlerServiceHandler is now stateful

### CODE REVIEW

**Comment by basilevs** [comment]
@akurtakov API checker complains about `MultiPageEditor` API. I had to bump org.eclipse.ui.workbench version. Is this the correct approach?

```
Error:  Failed to execute goal org.eclipse.tycho:tycho-apitools-plugin:4.0.4-SNAPSHOT:verify (verify) on project org.eclipse.ui.workbench: There are API errors:
Error:  META-INF/MANIFEST.MF:0 The type org.eclipse.ui.part.MultiPageEditor has been removed from org.eclipse.ui.workbench_3.132.0
Error:  Eclipse UI/org/eclipse/ui/activities/ActivityManagerEvent.java:235 Invalid @since 3.131 tag on copyFor(IActivityManager); expecting @since 3.132
Error:  Eclipse UI/org/eclipse/ui/activities/ActivityManagerEvent.java:210 Invalid @since 3.131 tag on ActivityManagerEvent(IActivityManager, boolean, boolean, boolean, Set<String>, Set<String>, Set<String>, Set<String>); expecting @since 3.132
Error:  Eclipse UI/org/eclipse/ui/activities/ActivityManagerEvent.java:321 Invalid @since 3.131 tag on haveEnabledNonExpressionControlledActivityIdsChanged(); expecting @since 3.132
Error:  Eclipse UI/org/eclipse/ui/activities/IActivityManager.java:123 Invalid @since 3.131 tag on dispose(); expecting @since 3.132
Error:  META-INF/MANIFEST.MF:5 The major version should be incremented in version 3.132.0, since API breakage occurred since version 3.130.0
Error:  -> [Help 1]
```

File: bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -2,7 +2,7 @@ Manifest-Version: 1.0
 Bundle-ManifestVersion: 2
 Bundle-Name: %pluginName
 Bundle-SymbolicName: org.eclipse.ui.workbench; singleton:=true
-Bundle-Version: 3.132.0.qualifier
+Bundle-Version: 4.0.0.qualifier
```
Comment (merks) [inline bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: This is a no go.  We can’t have broken api.

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: It is broken in master branch. See the commit about MultiPageEditor 5aefb9380169c779ff9c7e5a07a89a60d0c03135.
This PR does not break APIs (but does extend internal ones).

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: I'm investigating why master changes affect PR build's API checks in https://github.com/eclipse-platform/eclipse.platform.ui/pull/1254

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: Confirmed, a PR without any changes at all is failing due to previously broken API lacking version bump.

https://github.com/eclipse-platform/eclipse.platform.ui/pull/1254
https://github.com/eclipse-platform/eclipse.platform.ui/actions/runs/6677970750/job/18148533353?pr=1254

```
Error:  Failed to execute goal org.eclipse.tycho:tycho-apitools-plugin:4.0.4-SNAPSHOT:verify (verify) on project org.eclipse.ui.workbench: There are API errors:
Error:  META-INF/MANIFEST.MF:0 The type org.eclipse.ui.part.MultiPageEditor has been removed from org.eclipse.ui.workbench_3.131.100
Error:  META-INF/MANIFEST.MF:5 The major version should be incremented in version 3.131.100, since API breakage occurred since version 3.130.0
Error:  -> [Help 1]
```

Comment (merks) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: Very strange indeed. 😱

Comment (laeubi) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: If you look closer at the log you will see:

```
14:36:20.557 [WARNING] [API WARNING] File MANIFEST.MF at line 0: The API problem filter for: 'The major version should be incremented in version 3.131.0, since API breakage occurred since version 3.130.0' is no longer used (location: /home/jenkins/agent/workspace/eclipse.platform.ui_PR-1253/bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF)
14:36:20.557 [WARNING] [API WARNING] File MANIFEST.MF at line 0: The API problem filter for: 'The type org.eclipse.ui.part.MultiPageEditor has been removed from org.eclipse.ui.workbench_3.131.0' is no longer used (location: /home/jenkins/agent/workspace/eclipse.platform.ui_PR-1253/bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF)
```

so the API filter needs to be adjusted due to the changed version of the manifest

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: The argument here was, that verison should not have been changed at all. This PR does not have any API breakages.

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: Replaced major version bump with minor version bump and added another filter.

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: I've failed to bump minor version. If I do, IDE complains, that the bump is unnecessary. I suspect this is caused by the problem filter.

Comment (laeubi) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: Yes this is a nightmare, you need then another filter for the bump as well ...

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: > Yes this is a nightmare, you need then another filter for the bump as well ...

Adding filter makes version bump unnecessary
Removing filter reports API breakage for the bumped version

There is no way to bump version here anymore (and therefore, make any changes).

Comment (laeubi) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: You need one filter that suppress the major version bump, then you increase the minor and add another filter that suppress the "unnecessary" version bump error.

Comment (merks) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: I thought @laeubi suggested the filter needs updating to be applicable for the updated version?

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: Yes, when filter is updated, IDE complains, that the bump is unnecessary.

Comment (iloveeclipse) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: Why do you think you need a version bump at all?

**Comment by basilevs** [comment]
Now that Jenkins is happy (without version bump!) the test is failing:
```
Invalid test class 'org.eclipse.ui.tests.commands.StateTest':
  1. Test class should have exactly one public zero-argument constructor
	at org.junit.runners.ParentRunner.validate(ParentRunner.java:525)
	at org.junit.runners.ParentRunner.<init>(ParentRunner.java:102)
	at org.junit.runners.BlockJUnit4ClassRunner.<init>(BlockJUnit4ClassRunner.java:84)
	at org.junit.runners.JUnit4.<init>(JUnit4.java:23)
	at jdk.internal.reflect.GeneratedConstructorAccessor16.newInstance(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
	at org.junit.internal.builders.AnnotatedBuilder.buildRunner(AnnotatedBuilder.java:104)
	at org.junit.internal.builders.AnnotatedBuilder.runnerForClass(AnnotatedBuilder.java:86)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:70)
	at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:37)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:70)
	at org.junit.runners.model.RunnerBuilder.runners(RunnerBuilder.java:125)
	at org.junit.runners.model.RunnerBuilder.runners(RunnerBuilder.java:11
```

It was probably more compatible with Junit3.
UPDATE: migrated to Junit4

**Comment by basilevs** [comment]
Now that Jenkins is happy (without version bump!) the test is failing:
```
Invalid test class 'org.eclipse.ui.tests.commands.StateTest':
  1. Test class should have exactly one public zero-argument constructor
	at org.junit.runners.ParentRunner.validate(ParentRunner.java:525)
	at org.junit.runners.ParentRunner.<init>(ParentRunner.java:102)
	at org.junit.runners.BlockJUnit4ClassRunner.<init>(BlockJUnit4ClassRunner.java:84)
	at org.junit.runners.JUnit4.<init>(JUnit4.java:23)
	at jdk.internal.reflect.GeneratedConstructorAccessor16.newInstance(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
	at org.junit.internal.builders.AnnotatedBuilder.buildRunner(AnnotatedBuilder.java:104)
	at org.junit.internal.builders.AnnotatedBuilder.runnerForClass(AnnotatedBuilder.java:86)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:70)
	at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:37)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:70)
	at org.junit.runners.model.RunnerBuilder.runners(RunnerBuilder.java:125)
	at org.junit.runners.model.RunnerBuilder.runners(RunnerBuilder.java:11
```

It was probably more compatible with Junit3.

File: bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF (hunk 1 above)
Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: @iloveeclipse internal API in package `org.eclipse.ui.internal.handlers` is extended. It is shared with `org.eclipse.ui` and `org.eclipse.ui.ide`. Also, Jenkins complains if no version bump happens when code and qualifier are changed, but not version.

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: Anyway, I've removed the version bump (which should not be done, as API and code have changed) and the API verification is now passing. Maybe  @akurtakov can give some advice later, how to fix this.
Meanwhile, Jenkins produces test results so I can continue working on this draft.

Thanks for the help everyone!

Comment (iloveeclipse) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: There is no "internal" API  for API  tooling. So if **real** API isn't changed, **and** bundle version changed already during 4.30 cycle, no bump is needed.

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/META-INF/MANIFEST.MF]: > There is no "internal" API for API tooling.

@iloveeclipse thanks, I did not know that.
Exported package:
```
org.eclipse.ui.internal.handlers;x-friends:="org.eclipse.ui,org.eclipse.ui.ide",
```
looked "API enough" to me :sweat_smile:

Jenkins did not let me push other plugins without patch version bump, and I failed to realize their situation may be different from `org.eclipse.ui.workbench`.

**Comment by basilevs** [comment]
> It was probably more compatible with Junit3.

Migrated ti Junit4.

**Comment by basilevs** [comment]
HoverTest  seems to be unstable - passes locally and on Windows, fails on Linux.
```
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:87)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.junit.Assert.assertTrue(Assert.java:53)
	at org.eclipse.ui.genericeditor.tests.HoverTest.testProblemHover(HoverTest.java:130)
```
Enhanced the failure message
UPDATE: the test passed without meaningful  changes.
UPDATE2: the test failed again, save a screeshot to /home/runner/work/eclipse.platform.ui/eclipse.platform.ui/tests/org.eclipse.ui.genericeditor.tests/../../results/html/org.eclipse.ui.genericeditor.tests.HoverTest.testEnabledWhenHover.png, but this folder is not included in the test artifacts archive:
```
Run actions/upload-artifact@a8a3f3ad30e3422c9c7b888a15615d19a852ae32
  with:
    name: test-results-ubuntu-latest
    if-no-files-found: warn
    path: /home/runner/work/eclipse.platform.ui/eclipse.platform.ui/**/target/surefire-reports/*.xml
```
UPDATE3: https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/pull/1485

File: tests/org.eclipse.ui.tests/Eclipse UI Tests/org/eclipse/ui/tests/commands/StateTest.java [hunk 2]
```diff
@@ -146,10 +148,11 @@ protected final void doSetUp() {
 		final IHandlerService handlerService = fWorkbench
 				.getService(IHandlerService.class);
 		handlerActivation = handlerService.activateHandler(COMMAND_ID, handler);
+		command.setEnabled(handlerService.getCurrentState()); // Force handler initialization
```
Comment (basilevs) [inline tests/org.eclipse.ui.tests/Eclipse UI Tests/org/eclipse/ui/tests/commands/StateTest.java]: This is a problem. Original Eclipse was able to affect the command state immediately upon handler activaton. However, HandlerServiceHandler does not observe the currently active state of subhandlers activation and operates in pull mode.
Solution in this state works fine for menu items, but would probably fail to update toolbars.

UPDATE:
`org.eclipse.e4.ui.workbench.renderers.swt.ToolBarManagerRenderer.init()` subscribes to some context changes and calls `Command.setEnabled()` that might be enough. But a test would still be nice.

File: bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/handlers/HandlerProxy.java [hunk 3]
```diff
@@ -475,4 +481,20 @@ public String getAttributeName() {
 	public IHandler getHandler() {
 		return handler;
 	}
+
+	@Override
+	public void addState(String stateId, State state) {
+		super.addState(stateId, state);
```
Comment (basilevs) [inline bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/handlers/HandlerProxy.java:483]: Now that we handle state addition correctly, `handleStateChange()` should be done internally by handlers supporting both `IObjectWithState` and `IStateListener` protocols. However, as this functionality was broken for ten years, it is unlikely, that clients properly support the protocol. For this reason we are leaving the additional state changed event in `handleStateChanged()` as is and  clients that support both protocols (like `AbstractHandlerWithState`) will receive two events - "state added" and "state changed" instead of the original design, where "state added" was supposed to be enough.

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/handlers/HandlerProxy.java:483]: Removed `IStateListener` branch, as it could never be called due to bugs in `E4HandlerProxy`.

File: tests/org.eclipse.ui.tests/Eclipse UI Tests/org/eclipse/ui/tests/commands/StateTest.java (hunk 2 above)
Comment (basilevs) [reply tests/org.eclipse.ui.tests/Eclipse UI Tests/org/eclipse/ui/tests/commands/StateTest.java]: Added integration tests that ensure that toolbar item does indeed subscribe to command state and polls its handlers.

**Comment by basilevs** [comment]
AbstractTreeViewer tests are unstable, but other than that, this change seems to be ready for a review.

**Comment by basilevs** [comment]
AbstractTreeViewer tests are unstable, but other than that, this change seems to be ready for a review.

**Review by mickaelistria — APPROVED** [review]
This looks all good to me. If no one objects nor merge in the next days, I'll merge it. But anyone closer to the review is more than welcome to do it earlier.

**Comment by basilevs** [comment]
@sratz could you take a look?

**Comment by basilevs** [comment]
I've found potentially related ignored tests:
- CommandEnablementTest
- Bug66182Test
- ActionDelegateProxyTest

I probably need to ensure the design intent behind them is not is contradiction with my fixes.

UPDATE:
I've checked the tests. Those are indeed somewhat related to handler activation, but do not seem to be in direct conflict with my changes. I've created https://github.com/eclipse-platform/eclipse.platform.ui/issues/1261 that is detected by those tests and is particularly easy to fix.

File: bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/handlers/E4HandlerProxy.java [hunk 4]
```diff
@@ -50,7 +52,7 @@
  * @since 3.5
  *
  */
-public class E4HandlerProxy implements IHandler2, IHandlerListener, IElementUpdater {
+public class E4HandlerProxy implements IHandler2, IHandlerListener, IElementUpdater, IObjectWithState {
```
Comment (mickaelistria) [inline bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/handlers/E4HandlerProxy.java:55]: Since it is interpreted as an API breakage, requiring an API filter; couldn't we instead encapsulate the state as a package-visible final field ?

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/handlers/E4HandlerProxy.java:55]: The (internal) API is extended, not broken. I found that reducing visibility is impossible. Here is the chain of thought:

1. `org.eclipse.e4.core.commands.internal.HandlerServiceHandler` is used by Workbench to hydrate Command Framework
1. `HandlerServiceHandler` works with  `org.eclipse.e4.core.contexts.IEclipseContext`  available to consumers of Command Framework
1. Consumer can populate `IEclipseContext` with their own handlers using `org.eclipse.e4.core.commands.EHandlerService`
1. Consumer expects the protocols to match those used by Command and Workbench frameworks
    1. Command and Workbench frameworks use direct implementation of Java interface for various handlers protocols
1. HandlerServiceHandler has to detect the workbench protocols in the provided handlers to stay compatible with consumer's handlers. If other protocols are in use an additional different logic is required to detect them, which I deem unacceptable.
1. Therefore Workbench handlers have to use same protocols as consumer's handlers

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/handlers/E4HandlerProxy.java:55]: On a second thought, this is an internal API and @iloveeclipse says, API tools should not complain about extensions of internal APIs. I guess, I'll try to revert relevant version bumps.

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/handlers/E4HandlerProxy.java:55]: Relevant version bumps were reverted previously. `org.eclipse.ui.workbench` has original version, which indicates that API tools have not found any API changes at all (this is correct, as all changed classes are in internal packages available to `org.eclipse.ui` and `org.eclipse.ui.ide` plug-ins).

File: bundles/org.eclipse.ui.workbench/.settings/.api_filters [hunk 5]
```diff
@@ -4,7 +4,6 @@
         <filter comment="https://github.com/eclipse-platform/eclipse.platform.ui/issues/1155" id="923795461">
             <message_arguments>
                 <message_argument value="3.131.0"/>
-                <message_argument value="3.130.0"/>
```
Comment (mickaelistria) [inline bundles/org.eclipse.ui.workbench/.settings/.api_filters]: I still don't get why this change is necessary. What happens without it?

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/.settings/.api_filters]: Build fails with "The major version should be incremented in version 3.131.0, since API breakage occurred since version 3.130.0"
This happens due to MultiPageEditor removal in a previous push to master.
https://github.com/eclipse-platform/eclipse.platform.ui/pull/1253#discussion_r1375279272

Screenshot 2023-11-07 at 14 10 05

Comment (mickaelistria) [reply bundles/org.eclipse.ui.workbench/.settings/.api_filters]: Can't you just keep the current version in MANIFEST.MF? If I'm not mistaken, there is no externally visible change (all API changes are in internal code) and version was already bumped. So keeping same versions, and rebasing on master should not trigger new errors?

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/.settings/.api_filters]: The MANIFEST.MF for org.eclipse.ui.workbench and the version within are untouched in this MR. There are no conflicts in this MR, so rebase will have no effect.

Comment (mickaelistria) [reply bundles/org.eclipse.ui.workbench/.settings/.api_filters]: I've fetched the PR locally and reverted the .api_filters to previous version and see no error.
Can you please try to leave it unmodified? If build succeeds without modification there, then it means we've worried for nothing and we can move forward.

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/.settings/.api_filters]: I've force-pushed a revision without this change.

Comment (mickaelistria) [reply bundles/org.eclipse.ui.workbench/.settings/.api_filters]: Thanks, let's see...

Comment (basilevs) [reply bundles/org.eclipse.ui.workbench/.settings/.api_filters]: I've revisited the change history and now see, that it is a remnant from the attempt to bump workbench minor or patch versions. The bump is removed and the change might no longer be necessary.

**Comment by mickaelistria** [comment]
It's still OK to merge for M3, so let's merge.

File: bundles/org.eclipse.e4.core.commands/src/org/eclipse/e4/core/commands/internal/HandlerServiceHandler.java [hunk 6]
```diff
@@ -33,13 +35,15 @@
...
-public class HandlerServiceHandler extends AbstractHandler {
+public class HandlerServiceHandler extends AbstractHandlerWithState {
 
 	private static final String FAILED_TO_FIND_HANDLER_DURING_EXECUTION = "Failed to find handler during execution"; //$NON-NLS-1$
 	private static final String HANDLER_MISSING_EXECUTE_ANNOTATION = " handler is missing @Execute"; //$NON-NLS-1$
 	private static final Object missingExecute = new Object();
 
 	protected final String commandId;
+	// Remove state from currentStateHandler when it goes out of scope
+	private IObjectWithState currentStateHandler;
```
Comment (iloveeclipse) [inline bundles/org.eclipse.e4.core.commands/src/org/eclipse/e4/core/commands/internal/HandlerServiceHandler.java:46]: This field leaks whole editor instances, see new 4 test fails:

https://download.eclipse.org/eclipse/downloads/drops4/I20231107-1800/testresults/html/org.eclipse.jdt.ui.tests_ep430I-unit-cen64-gtk3-java17_linux.gtk.x86_64_17.html

Please provide fix ASAP.

Comment (basilevs) [reply bundles/org.eclipse.e4.core.commands/src/org/eclipse/e4/core/commands/internal/HandlerServiceHandler.java:46]: Quick fix is impossible, as the field was an important part of the original PR. We have to revert the whole thing.
@mickaelistria could you please do so?

Comment (laeubi) [reply bundles/org.eclipse.e4.core.commands/src/org/eclipse/e4/core/commands/internal/HandlerServiceHandler.java:46]: If its really a leak, why not use a `WeakReference` here?

Comment (basilevs) [reply bundles/org.eclipse.e4.core.commands/src/org/eclipse/e4/core/commands/internal/HandlerServiceHandler.java:46]: @laeubi yes, weak reference might work. I will attempt to apply it now.

Comment (iloveeclipse) [reply bundles/org.eclipse.e4.core.commands/src/org/eclipse/e4/core/commands/internal/HandlerServiceHandler.java:46]: Have no IDE at hand, but isn't possible to clear the reference on dispose()?

Comment (basilevs) [reply bundles/org.eclipse.e4.core.commands/src/org/eclipse/e4/core/commands/internal/HandlerServiceHandler.java:46]: @iloveeclipse the reference is cleared on dispose in the original PR. Unfortunately, most of these handlers have lifetime of Workbench.

Comment (basilevs) [reply bundles/org.eclipse.e4.core.commands/src/org/eclipse/e4/core/commands/internal/HandlerServiceHandler.java:46]: Created https://github.com/eclipse-platform/eclipse.platform.ui/pull/1291 with `WeakReference`. There is a problem with it (objects are collected without removal of listeners), but it does not leak.
The test for leaks would be hard to write.
